### PR TITLE
[Doc] Document multi-agent _step TensorDict layout

### LINKED ISSUE #2425: Document the TensorDict structure of the return of the _step() function for a multi agent environment
## Motivation

It is not very clear what should be the structure of TensorDict of the return of the _step() function for a multi agent environment.

If there are two agents A and B and they are in separate groups, what would be the structure of the TensorDict that is returned by the _step() function?

## Solution

Update the documentation and provide an example of a multi agent environment that is written natively in torchrl and it is not translated from other frameworks like petting zoo  

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

@matteobettini

### PR DESCRIPTION
**Pull request by YeonwooSung**
## Description

Extends `docs/source/reference/envs_multiagent.rst` with a **Multiple agent groups** section so a native `EnvBase._step()` is documented beyond the single `"agents"` group used by VMAS.

The new section:

- Shows the key tree a custom `_step()` must write for several groups (`"red"` / `"blue"`, or `"agents"` / `"adversaries"`): per-group nested tensordicts with a stacked agent dimension.
- States where **action** (input only), **observation**, and **reward** live, and where **done** / **terminated** / **truncated** live when they are shared (root reset signal), per-group / per-agent (stacked), or one-group-per-agent.
- Requires specs to mirror that nesting, with a two-group spec example and an optional per-group done spec.
- Includes a copy-paste native `TwoTeamEnv(EnvBase)` sketch (not a PettingZoo wrapper) that reads `(group, "action")` and returns observations, rewards, and shared done flags without a `"next"` wrapper.
- Cross-links the [competitive MADDPG tutorial](https://pytorch.org/rl/stable/tutorials/multiagent_competitive_ddpg.html) as the canonical multi-group training loop.

The existing VMAS single-group example is kept as the simple case. The intro now treats `"agents"` as one group name, not a required key.

### Code example

Example return value from a custom `_step`: two groups with different agent counts and shared environment termination. EnvBase.step adds the outer `"next"` key.

```python
import torch
from tensordict import TensorDict

batch_size = 4
out = TensorDict({
    group: TensorDict({
        "observation": torch.zeros(batch_size, n_agents, 8),
        "reward": torch.zeros(batch_size, n_agents, 1),
    }, batch_size=[batch_size, n_agents])
    for group, n_agents in {"red": 2, "blue": 3}.items()
}, batch_size=[batch_size])
out["done"] = torch.zeros(batch_size, 1, dtype=torch.bool)
out["terminated"] = torch.zeros_like(out["done"])
assert out["red", "reward"].shape == (4, 2, 1)
assert out["blue", "observation"].shape == (4, 3, 8)
# return out from _step; group actions are read from the input TensorDict.
```

## Motivation and Context

close #2425

The current multi-agent env page only documented the single `"agents"` group. Matteo Bettini agreed it needed a more general multi-group description; a later comment pointed at the competitive DDPG tutorial rollout as the pattern.

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

## Types of changes

- [x] Documentation (update in the documentation)

## Checklist

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/source/reference/envs_multiagent.rst [hunk 1]
```diff
@@ -130,6 +139,358 @@ single agent standards.
...
+- **Reward.** Output of ``_step()`` only (not ``_reset()``). Per-agent or
+  per-group rewards live at ``(group, "reward")`` with shape
+  ``(*batch, n_agents_in_group, 1)``. A fully shared reward is a root
+  ``"reward"`` with shape ``(*batch, 1)``.
+- **Done / terminated / truncated** can sit at three levels, which can be
+  combined:
+
+  * **Shared (root).** ``"done"``, ``"terminated"`` and (if used)
+    ``"truncated"`` at the root, shape ``(*batch, 1)``. This is the signal
+    TorchRL uses to reset the environment. A native env must write at least
```
Comment (vmoens) [inline docs/source/reference/envs_multiagent.rst]: [P2] Do not require a root done flag for every native env

EnvBase also supports environments whose done/terminated/reset keys are entirely nested; NestedCountingEnv(nest_done=True) and the existing nested StepCounter tests exercise that contract. Root shared flags are the convention for this example and some wrappers, not a universal native-env requirement. Qualify this paragraph so implementers are not told to add a second termination hierarchy unnecessarily.

File: docs/source/reference/envs_multiagent.rst [hunk 2]
```diff
@@ -130,6 +139,358 @@ single agent standards.
...
+    this shared flag (VMAS writes only this).
+  * **Per-group / per-agent, stacked.** ``(group, "done")`` (and the
+    ``terminated`` / ``truncated`` siblings) with shape
+    ``(*batch, n_agents_in_group, 1)``. PettingZoo writes these *and*
+    aggregates them into the root flag (``any`` or ``all``, controlled by
+    ``done_on_any``).
+  * **One group per agent.** With
+    :attr:`~torchrl.envs.MarlGroupMapType.ONE_GROUP_PER_AGENT` each group
+    has a single agent, so ``(agent_name, "done")`` has shape
+    ``(*batch, 1)`` and there is no extra agent dimension.
```
Comment (vmoens) [inline docs/source/reference/envs_multiagent.rst]: [P2] Keep the singleton agent dimension for one-agent groups

ONE_GROUP_PER_AGENT changes group membership, not the group TensorDict layout. PettingZoo constructs each group Composite with shape [n_agents] and each done leaf with [n_agents, 1], so a one-agent group still has trailing shape [1, 1] (or [*batch, 1, 1] with an env batch). Saying there is no agent dimension will make readers build incompatible specs/policies. Retain that singleton dimension in the documented shape.

**Review by vmoens — CHANGES_REQUESTED** [review]
The TwoTeamEnv example runs and passes check_env_specs. Two statements in the new layout contract do not match the supported environment/spec behavior.

File: docs/source/reference/envs_multiagent.rst (hunk 2 above)
Comment (YeonwooSung) [reply docs/source/reference/envs_multiagent.rst]: Addressed in 7ac9875cc. ONE_GROUP_PER_AGENT is documented as a grouping change only. PettingZoo still builds each group Composite with shape [n_agents] and each done leaf with [n_agents, 1], so a one-agent group keeps trailing shape [*batch, 1, 1]. The previous "no extra agent dimension" wording is gone.

File: docs/source/reference/envs_multiagent.rst (hunk 1 above)
Comment (YeonwooSung) [reply docs/source/reference/envs_multiagent.rst]: Addressed in 7ac9875cc. The paragraph now states that root done/terminated/truncated is the convention used by the TwoTeamEnv sketch and wrappers such as VMAS, not a native-env requirement. EnvBase already resets from nested done/_reset keys (NestedCountingEnv(nest_done=True, has_root_done=False) and the nested StepCounter tests), so the text no longer tells implementers to add a second termination hierarchy.

**Comment by YeonwooSung** [comment]
Updated the layout contract in 7ac9875cc for the two CHANGES_REQUESTED items:

- Root done/terminated/truncated is documented as a convention (TwoTeamEnv, VMAS), not a native-env requirement. EnvBase already supports fully nested done/_reset keys.
- ONE_GROUP_PER_AGENT keeps the singleton agent dimension: Composite shape `(*batch, 1)`, done leaf `(*batch, 1, 1)`.

The TwoTeamEnv sketch is unchanged and still uses shared root done flags.

**Review by vmoens — APPROVED** [review]
Re-reviewed 7ac9875. The text now allows nested-only done specs instead of requiring a root done universally, and the one-group-per-agent example retains the required singleton done dimension.

I reran the TwoTeamEnv example and check_env_specs successfully. The revised examples and shape descriptions are consistent. No remaining actionables.
